js/src/edit-free-campaign/hasUnsavedShippingRages.js

```diff
@@ -0,0 +1,43 @@
+/**
+ * External dependencies
+ */
+import { isEqual, differenceWith, at } from 'lodash';
+
+/**
+ * @typedef {import('.~/data/actions').ShippingRate} ShippingRate
+ */
+
+/**
+ * This function compares the key fields of two shipping rate arrays
+ * and returns whether there are any unsaved shipping changes.
+ *
+ * - Order:     The order of two arrays does not need to be the same.
+ * - ID:        The `id` is not compared.
+ *              On the UI, if a rate was deleted and then added,
+ *              then the `id` would not exist.
+ * - Threshold: If `options` or `options.free_shipping_threshold` does not exist,
+ *              `free_shipping_threshold` is treated as undefined.
+ *
+ * @param  {Array<ShippingRate>} rates The shipping rates to be compared.
+ * @param  {Array<ShippingRate>} savedRates The saved shipping rates received from API.
+ * @return {boolean} Whether there are any unsaved shipping changes.
+ */
+export default function hasUnsavedShippingRages( rates, savedRates ) {
+	if ( rates.length !== savedRates.length ) {
+		return true;
+	}
+
+	const paths = [
+		'country',
+		'method',
+		'currency',
+		'rate',
+		'options.free_shipping_threshold',
+	];
+
+	const diffRates = differenceWith( rates, savedRates, ( a, b ) => {
+		return isEqual( at( a, paths ), at( b, paths ) );
```

Comment:
💅 Just a related thought that I had when I was thinking about this issue, we could probably refactor this into a new function like `isEqualShippingRate (a, b)`, then `paths` will be inside that `isEqualShippingRate` function, and then this line will be just `return isEqualShippingRate( a, b );`.

Note: we also already have a function called `isSameShippingRate`. Thinking we might also get into some confusions later. 😅

Reply:
Thanks for the idea!

I lean toward extracting this when it needs to be reused, then it would not be confused with `isSameShippingRate` before that.